internal/transport/http_util.go

```diff
@@ -244,51 +266,112 @@ func (d *decodeState) decodeHeader(frame *http2.MetaHeadersFrame) error {
 	if frame.Truncated {
 		return status.Error(codes.Internal, "peer header list size exceeded limit")
 	}
+
+	// isGRPC indicates whether the peer is speaking gRPC (otherwise HTTP).
+	//
+	// We are in gRPC mode (peer speaking gRPC) if:
+	// 	* We are client side and have already received a HEADER frame that indicates gRPC peer.
+	//  * The header contains valid  a content-type, i.e. a string starts with "application/grpc"
+	// And we should handle error specific to gRPC.
+	//
+	// Otherwise (i.e. a content-type string starts without "application/grpc", or does not exist), we
+	// are in HTTP fallback mode, and should handle error specific to HTTP.
+	//
+	// d.ignoreContentType is only set on client side after a gRPC ResponseHeader has been received (indicating
+	// peer speaking gRPC). Therefore, we can initialized isGRPC to d.ignoreContentType.
+	isGRPC := d.ignoreContentType
+	var grpcErr, httpErr, contentTypeErr error
 	for _, hf := range frame.Fields {
+		if hf.Name != "content-type" && hf.Name != ":status" && grpcErr != nil {
+			// if we've already encountered grpc related field parsing error, then we skip processing
+			// all following grpc related field.
+			continue
+		}
+
 		if err := d.processHeaderField(hf); err != nil {
-			return err
+			switch hf.Name {
+			case "content-type":
+				contentTypeErr = err
+			case ":status":
+				httpErr = err // In gRPC mode, we don't care about HTTP field parsing error, so we store it separately.
+			default:
+				grpcErr = err // store the first encountered gRPC field parsing error.
+			}
+			continue
 		}
-	}
 
-	if d.serverSide {
-		return nil
+		// we got a valid content-type that starts with "applicatin/grpc", so we are operating in grpc
+		// mode.
+		if hf.Name == "content-type" {
+			isGRPC = true
```

Comment:
It may be cleaner to put "isGRPC" into parsedHeaderData to avoid setting it here based on knowledge of how processHeaderField works (i.e. no error and field name = "content-type" implies GRPC).

Maybe the above logic can all be moved there too?  (The switch and the knowledge of skipping certain fields.)

Reply:
done